### PR TITLE
[None][perf] Integrate the flashinfer gdn prefill kernel for qwen3.5

### PR DESCRIPTION
**Pull request by nv-guomingz**
## Summary by CodeRabbit

* **New Features**
  * Added FlashInfer integration for optimized GDN (Gated Delta Network) prefill operations, with optional environment variable configuration to control which implementation is used.
  * Configurable fallback to existing implementation for systems without FlashInfer support.

* **Tests**
  * Added operator-level parity tests validating FlashInfer GDN adapter against reference implementation across various sequence configurations.
  * Added environment variable routing tests to verify correct implementation selection based on configuration.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by xinhe-nv — APPROVED** [review]

File: tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    initial_state_indices: Optional[torch.Tensor] = None,
+    inplace_indexed_state_update: bool = False,
+    output_final_state: bool = False,
+    cu_seqlens: Optional[torch.Tensor] = None,
+    head_first: bool = False,
+    use_qk_l2norm_in_kernel: bool = False,
+) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+    """Adapter for FlashInfer's chunk_gated_delta_rule.
+
+    Args/returns: see ``docs/superpowers/specs/2026-04-29-flashinfer-gdn-prefill-design.md``.
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py]: remove this line

File: tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    # --- Step 1: pre-flight asserts --------------------------------------
+    assert head_first is False, "head_first=True is not supported by this wrapper"
+    assert q.dim() == 4 and q.shape[0] == 1, f"q must be [1, T, H_q, D_k], got {tuple(q.shape)}"
+    assert k.shape[2] == q.shape[2], (
+        f"num_q_heads ({q.shape[2]}) must equal num_k_heads ({k.shape[2]})"
+    )
+    assert q.dtype in (torch.bfloat16, torch.float16), f"q dtype must be bf16/fp16, got {q.dtype}"
+    assert g.dtype == torch.float32, f"g must be fp32, got {g.dtype}"
+    assert cu_seqlens is not None, "cu_seqlens is required (varlen mode)"
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py:73]: would it support padding mode?

File: tensorrt_llm/_torch/modules/mamba/gdn_mixer.py [hunk 3]
```diff
@@ -3,6 +3,7 @@
 # SPDX-FileCopyrightText: Copyright (c) 2022-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 # SPDX-License-Identifier: Apache-2.0
 
+import os as _os
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/mamba/gdn_mixer.py]: curious why we need this line?

File: tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    confirmed by ``flashinfer/tests/gdn/test_prefill_delta_rule.py`` which
+    transposes the last two dims to compare with the reference implementation).
+    The wrapper transposes initial_state on the way in and output_state on the
+    way out so callers see TRT-LLM's ``[N, H, K, V]`` everywhere.
+
+This module is only imported when ``TLLM_USE_FLASHINFER_GDN_PREFILL=1`` is set
+at process start; do not import it lazily inside hot paths.
+"""
+
+from __future__ import annotations
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py]: curious why this line is needed

**Review by Wanli-Jiang — APPROVED** [review]

File: tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py (hunk 1 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py]: Done

File: tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py (hunk 2 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py:73]: No — this wrapper is varlen-only by design. All TRT-LLM call sites pass packed [1, T_total, H, D] + cu_seqlens (in-flight batching), and FlashInfer's underlying kernel only accepts varlen layout. If a future caller ever needs [B, T, H, D] padding mode, the trivial path is to repack as [1, B*T, H, D] with cu_seqlens=[0, T, 2T, …, B*T] and reuse the same code — but there's no such caller today, so we explicitly assert it out instead of adding unreachable code.

File: tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py (hunk 4 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py]: Removed.

File: tensorrt_llm/_torch/modules/mamba/gdn_mixer.py (hunk 3 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/modules/mamba/gdn_mixer.py]: The intention is to avoid calling like `from gdn_mixer import *`, but it seems  it won't happen. Update to `os` for more simplicity.

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 5]
```diff
@@ -6491,12 +6479,14 @@ def test_bf16_mtp(self, mtp_flag, mocker):
...
         model_dir = f"{self.MODEL_PATH}-FP8"
+
         # Model is being added to CI. Skip at the moment.
         if not os.path.exists(model_dir):
             pytest.skip(f"Model directory {model_dir} does not exist")
 
         kv_cache_config = KvCacheConfig(free_gpu_memory_fraction=0.8,
-                                        enable_block_reuse=enable_block_reuse)
+                                        enable_block_reuse=enable_block_reuse,
+                                        mamba_state_cache_interval=1024)
```
Comment (yuxianq) [inline tests/integration/defs/accuracy/test_llm_api_pytorch.py]: Why we need to change `mamba_state_cache_interval`?

File: tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    cu_seqlens: Optional[torch.Tensor] = None,
+    head_first: bool = False,
+    use_qk_l2norm_in_kernel: bool = False,
+) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+    """Adapter for FlashInfer's chunk_gated_delta_rule."""
+    # FlashInfer is imported lazily so importing this module on a non-FlashInfer
+    # build does not error until the function is actually called.
+    import flashinfer
+
+    # First-call signpost on every rank (once per process). Uses bare ``print``
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py]: It seems that you do not use any print here, is this comment outdated?

File: tests/unittest/_torch/modules/mamba/test_flashinfer_chunk_gdn.py [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+        head_first=False,
+        use_qk_l2norm_in_kernel=True,
+    )
+
+    torch.testing.assert_close(out_fi, out_triton, atol=2e-2, rtol=2e-2)
+
+
+@skip_unsupported
+def test_packed_initial_state_with_output_final_state_matches_triton():
+    """Site 2 (line 649) shape: already-packed initial state, output_final_state=True."""
```
Comment (yuxianq) [inline tests/unittest/_torch/modules/mamba/test_flashinfer_chunk_gdn.py]: What does `Site 2 (line 649)` mean?

File: tests/unittest/_torch/modules/mamba/test_flashinfer_chunk_gdn.py [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+        final_fi.to(torch.float32),
+        final_triton.to(torch.float32),
+        atol=5e-2,
+        rtol=5e-2,
+    )
+
+
+@skip_unsupported
+def test_indexed_gather_inplace_scatter_matches_triton():
+    """Site 1 (line 713) shape: full SSM pool + indices + inplace writeback."""
```
Comment (yuxianq) [inline tests/unittest/_torch/modules/mamba/test_flashinfer_chunk_gdn.py]: What does `Site 1 (line 713)` mean?

File: tests/unittest/_torch/modules/mamba/test_flashinfer_chunk_gdn.py [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+    reason="FlashInfer GDN prefill requires SM90 (Hopper) or SM100 (Blackwell)",
+)
+
+
+# Input factory ------------------------------------------------------------
+
+
+@torch.no_grad()
+def _make_inputs(
+    seq_lens: List[int],
```
Comment (yuxianq) [inline tests/unittest/_torch/modules/mamba/test_flashinfer_chunk_gdn.py]: We prefer `list[int]` now, can you update all `List` type in this file?

File: tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    # --- Step 5+6: call FlashInfer with pre-allocated output/state buffers
+    # FI 0.6.10 accepts `output=` / `output_state=`; pre-allocating skips its
+    # internal `torch.empty` per call. `num_o_heads` per FI docstring is
+    # max(num_q_heads, num_v_heads) — equivalently num_v_heads for GVA.
+    total_seq_len = q3.shape[0]
+    num_o_heads = max(q3.shape[1], v3.shape[1])
+    head_size = q3.shape[2]
+    num_seqs = cu_seqlens.shape[0] - 1
+    output_buf = q3.new_empty(total_seq_len, num_o_heads, head_size)
+    state_buf = q3.new_empty(num_seqs, num_o_heads, head_size, head_size, dtype=torch.float32)
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py]: We allocate state_buf even when we don't use it (both output_final_state and inplace_indexed_state_update are False), should we make it optional?

**Review by HuiGao-NV — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py (hunk 6 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py]: Yes, this comment should be removed

File: tests/unittest/_torch/modules/mamba/test_flashinfer_chunk_gdn.py (hunk 9 above)
Comment (nv-guomingz) [reply tests/unittest/_torch/modules/mamba/test_flashinfer_chunk_gdn.py]: Updated

File: tests/unittest/_torch/modules/mamba/test_flashinfer_chunk_gdn.py (hunk 7 above)
Comment (nv-guomingz) [reply tests/unittest/_torch/modules/mamba/test_flashinfer_chunk_gdn.py]: Update the comments. The original word site refers to different scenarios.

File: tests/unittest/_torch/modules/mamba/test_flashinfer_chunk_gdn.py (hunk 8 above)
Comment (nv-guomingz) [reply tests/unittest/_torch/modules/mamba/test_flashinfer_chunk_gdn.py]: Updated

File: tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py (hunk 10 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/modules/fla/flashinfer_chunk.py]: Updated

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py (hunk 5 above)
Comment (nv-guomingz) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py]: this case had been [waived](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tests/integration/test_lists/waives.txt#L150C96-L150C118) in https://nvbugs/6210714.
My patch is WAR when I filed the PR.

**Review by yuxianq — APPROVED** [review]

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py (hunk 5 above)
Comment (yuxianq) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py]: After offline discussion, remove this WAR from this PR, thanks~
